src/lib/libc/time/ctime.3
.\"	$MirOS: src/lib/libc/time/ctime.3,v 1.3 2005/09/22 20:33:01 tg Exp $
.\"	$OpenBSD: ctime.3,v 1.31 2005/07/22 03:23:37 jaredy Exp $
.\"
.\"
.Dd $Mdocdate: August 8 2007 $
.Dt CTIME 3
.Os
.Sh NAME
.Nm asctime ,
.Nm asctime_r ,
.Nm ctime ,
.Nm ctime_r ,
.Nm difftime ,
.Nm gmtime ,
.Nm gmtime_r ,
.Nm localtime ,
.Nm localtime_r ,
.Nm mktime ,
.Nm timegm ,
.Nm timelocal
.Nd convert date and time to ASCII
.Sh SYNOPSIS
.Fd #include <sys/types.h>
.Fd #include <time.h>
.Pp
.Fd extern char *tzname[2];
.Ft void
.Fn tzset "void"
.Ft "char *"
.Fn ctime "const time_t *clock"
.Ft "char *"
.Fn ctime_r "const time_t *clock" "char *buf"
.Ft double
.Fn difftime "time_t time1" "time_t time0"
.Ft "char *"
.Fn asctime "const struct tm *tm"
.Ft "char *"
.Fn asctime_r "const struct tm *tm" "char *buf"
.Ft "struct tm *"
.Fn localtime "const time_t *clock"
.Ft "struct tm *"
.Fn localtime_r "const time_t *clock" "struct tm *result"
.Ft "struct tm *"
.Fn gmtime "const time_t *clock"
.Ft "struct tm *"
.Fn gmtime_r "const time_t *clock" "struct tm *result"
.Ft time_t
.Fn mktime "struct tm *tm"
.Ft time_t
.Fn timegm "struct tm *tm"
.Ft time_t
.Fn timelocal "struct tm *tm"
.Sh DESCRIPTION
The
.Fn ctime
function converts a
.Li time_t ,
pointed to by
.Fa clock ,
representing the time in seconds since
00:00:00 UTC, 1970-01-01,
and returns a pointer to a
string of the form
.Pp
.Dl Thu Nov 24 18:22:48 1986\en
.Pp
Years requiring fewer than four characters are padded with leading zeroes.
For years longer than four characters, the string is of the form
.Pp
.Dl Thu Nov 24 18:22:48\ \ \ \ \ 81986\en
.Pp
with five spaces before the year.
These unusual formats are designed to make it less likely that older
software that expects exactly 26 bytes of output will mistakenly output
misleading values for out-of-range years.
.Pp
The
.Fn ctime_r
function converts the calendar time pointed to by
.Fa clock
to local time in exactly the same way as
.Fn ctime
and puts the string into the array pointed to by
.Fa buf
(which contains at least 26 bytes) and returns
.Fa buf .
Unlike
.Fn ctime ,
the thread-safe version
.Fn ctime_r
is not required to set
.Fa tzname .
.Pp
The
.Fn localtime
and
.Fn gmtime
functions return pointers to
.Li tm
structures, described below.
.Fn localtime
corrects for the time zone and any time zone adjustments
(such as Daylight Saving Time in the United States).
After filling in the
.Li tm
structure,
.Fn localtime
sets the
.Fa tm_isdst Ns 'th
element of
.Fa tzname
to a pointer to an
ASCII string that's the time zone abbreviation to be used with
the return value of
.Fn localtime .
.Pp
.Fn gmtime
converts to Coordinated Universal Time.
.Pp
The
.Fn localtime_r
and
.Fn gmtime_r
functions convert the calendar time pointed to by
.Fa clock
into a broken-down time in exactly the same way as their non-reentrant
counterparts,
.Fn localtime
and
.Fn gmtime ,
but instead store the result directly into the structure pointed to by
.Fa result .
Unlike
.Fn localtime ,
the reentrant version is not required to set
.Fa tzname .
.Pp
.Fn asctime
converts a time value contained in a
.Li tm
structure to a string,
as shown in the above example,
and returns a pointer to the string.
.Fn asctime_r
uses the buffer pointed to by
.Fa buf
(which should contain at least 26 bytes) and then
returns
.Fa buf .
.Pp
.Fn mktime
converts the broken-down time,
expressed as local time,
in the structure pointed to by
.Fa tm
into a calendar time value with the same encoding as that of the values
returned by the
.Fn time
function.
The original values of the
.Fa tm_wday
and
.Fa tm_yday
components of the structure are ignored,
and the original values of the other components are not restricted
to their normal ranges.
(A positive or zero value for
.Fa tm_isdst
causes
.Fn mktime
to presume initially that summer time (for example, Daylight Saving Time
in the U.S.A.)\&
respectively,
is or is not in effect for the specified time.
A negative value for
.Fa tm_isdst
causes the
.Fn mktime
function to attempt to divine whether summer time is in effect
for the specified time.)
On successful completion, the values of the
.Fa tm_wday
and
.Fa tm_yday
components of the structure are set appropriately,
and the other components are set to represent the specified calendar time,
but with their values forced to their normal ranges; the final value of
.Fa tm_mday
is not set until
.Fa tm_mon
and
.Fa tm_year
are determined.
.Fn mktime
returns the specified calendar time;
if the calendar time cannot be represented,
it returns \-1.
.Pp
.Fn timelocal
is a deprecated interface that is equivalent to calling
.Fn mktime
with a negative value for
.Fa tm_isdst .
.Pp
.Fn timegm
is a deprecated interface that converts the broken-down time, as returned by
.Fn gmtime ,
into a calendar time value with the same encoding as that of the values
returned by the
.Fn time
function.
.Pp
.Fn difftime
returns the difference between two calendar times,
.Pf ( Fa time1 No \- Fa time0 ) ,
expressed in seconds.
.Pp
Declarations of all the functions and externals, and the
.Li tm
structure, are in the
.Aq Pa time.h
header file.
The structure (of type)
.Li struct tm
includes the following fields:
.Bd -literal -offset indent
int tm_sec;		/* seconds [0\-60] */
int tm_min;		/* minutes [0\-59] */
int tm_hour;		/* hours [0\-23] */
int tm_mday;		/* day of month [1\-31] */
int tm_mon;		/* month of year \- 1 [0\-11] */
time_t tm_year;		/* years since 1900 */
int tm_wday;		/* day of week (0 = sunday) */
int tm_yday;		/* day of year [0\-365] */
int tm_isdst;		/* summer time? [0 no, \*(Gt0 yes, \*(Lt0 unknown] */
long tm_gmtoff;		/* offset from UTC in seconds */
char *tm_zone;		/* abbreviation of timezone name */
.Ed
.Pp
The
.Fa tm_zone
and
.Fa tm_gmtoff
fields exist, and are filled in, only if arrangements to do
so were made when the library containing these functions was
created.
There is no guarantee that these fields will continue to exist
in this form in future releases of this code.
.Pp
.Fa tm_isdst
is greater than zero if summer time is in effect.
.Pp
.Fa tm_gmtoff
is the offset (in seconds) of the time represented
from UTC, with positive values indicating east
of the Prime Meridian.
.Sh FILES
.Bl -tag -width "/usr/share/zoneinfo/posixrules" -compact
.It Pa /usr/share/zoneinfo
time zone information directory
.It Pa /etc/localtime
local time zone file
.It Pa /usr/share/zoneinfo/posixrules
used with POSIX-style TZ's
.It Pa /usr/share/zoneinfo/UTC
for UTC leap seconds
.El
.Pp
If
.Pa /usr/share/zoneinfo/UTC
is absent,
UTC leap seconds are loaded from
.Pa /usr/share/zoneinfo/posixrules .
.Sh SEE ALSO
.Xr getenv 3 ,
.Xr strftime 3 ,
.Xr time 3 ,
.Xr tzset 3 ,
.Xr tzfile 5 ,
.Xr zic 8
.Sh NOTES
The return values
of the non re-entrant functions
point to static data;
the data is overwritten by each call.
The
.Fa tm_zone
field of a returned
.Li struct tm
points to a static array of characters, which
will also be overwritten at the next call
(and by calls to
.Fn tzset ) .
.Pp
.Fn asctime
and
.Fn ctime
behave strangely for years before 1000 or after 9999.
The 1989 and 1999 editions of the C Standard say
that years from \-99 through 999 are converted without
extra spaces, but this conflicts with longstanding
tradition and with this implementation.
Traditional implementations of these two functions are
restricted to years in the range 1900 through 2099.
To avoid this portability mess, new programs should use
.Fn strftime
instead.
.Pp
The default system time zone may be set by running
.Dq Li zic -l timezone
as the superuser.
.Pp
Avoid using out-of-range values with
.Fn mktime
when setting up lunch with promptness sticklers in Riyadh.
.\" based on @(#)newctime.3	7.16
.\" This file is in the public domain, so clarified as of
.\" 2003-02-14 by Arthur David Olson (arthur_david_olson@nih.gov).


src/sys/lib/libsa/ctime.c
/*	$OpenBSD: ctime.c,v 1.5 2003/08/11 06:23:09 deraadt Exp $	*/

/*
 * Copyright (c) 1998 Michael Shalayeff
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/time.h>
#include "stand.h"

#define isleap(y) ((((y) % 4) == 0 && ((y) % 100) != 0) || ((y) % 400) == 0)

char *
ctime(const time_t *clock)
{
	static const char wdays[][4] = {
		"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"
	};
	static const char months[][4] = {
		"Jan", "Feb", "Mar", "Apr", "May", "Jun",
		"Jul", "Aug", "Sep", "Oct", "Nov", "Dec"
	};
	static const u_int monthcnt[] = {
		31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31
	};
	static char buf[32];
	int ss, mm, hh, wday, month, year;
	time_t tt = *clock;

	ss = tt % 60;
	tt /= 60;	/* minutes */
	mm = tt % 60;
	tt /= 60;	/* hours */
	hh = tt % 24;
	tt /= 24;	/* days */
	wday = (4 + tt) % 7;	/* weekday, 'twas thursday when time started */

	for (year = 1970; tt >= 365; year++)
		tt -= isleap(year)? 366: 365;

	tt++;	/* days are 1-based */

	for (month = 0; tt > monthcnt[month]; month++)
		tt -= monthcnt[month];

	if (month > 2 && isleap(year))
		tt--;

	/* no field widths in printf() */
	snprintf(buf, sizeof buf, "%s %s %d %d:%d:%d %d\n",
	    ((wday  < 0 || wday  >=  7)? "???": wdays[wday]),
	    ((month < 0 || month >= 12)? "???": months[month]),
	    (int)tt, hh, mm, ss, year);
	return buf;
}
